core/io/networked_multiplayer_custom.cpp

```diff
@@ -0,0 +1,184 @@
+
+Error NetworkedMultiplayerCustom::get_packet(const uint8_t **r_buffer, int &r_buffer_size) {
+	ERR_FAIL_COND_V(incoming_packets.size() == 0, Error::ERR_UNAVAILABLE);
+
+	current_packet = incoming_packets.front()->get();
+	incoming_packets.pop_front();
+
+	*r_buffer = current_packet.data.read().ptr();
+	r_buffer_size = current_packet.data.size();
+
+	return Error::OK;
+}
+
+Error NetworkedMultiplayerCustom::put_packet(const uint8_t *p_buffer, int p_buffer_size) {
+	PoolByteArray a;
+	a.resize(p_buffer_size);
+
+	PoolByteArray::Write w = a.write();
+	memcpy(w.ptr(), p_buffer, p_buffer_size);
+
+	emit_signal("packet_generated", target_id, a, transfer_mode);
+
+	return Error::OK;
+}
+
+int NetworkedMultiplayerCustom::get_available_packet_count() const {
+	return incoming_packets.size();
+}
+
+int NetworkedMultiplayerCustom::get_max_packet_size() const {
+	return max_packet_size;
+}
+
+//
+// NetworkedMultiplayerPeer.
+//
+
+void NetworkedMultiplayerCustom::set_transfer_mode(NetworkedMultiplayerPeer::TransferMode p_mode) {
+	transfer_mode = p_mode;
+}
+
+NetworkedMultiplayerPeer::TransferMode NetworkedMultiplayerCustom::get_transfer_mode() const {
+	return transfer_mode;
+}
+
+void NetworkedMultiplayerCustom::set_target_peer(int p_target_peer) {
+	target_id = p_target_peer;
+}
+
+int NetworkedMultiplayerCustom::get_packet_peer() const {
+	ERR_FAIL_COND_V(connection_status != ConnectionStatus::CONNECTION_CONNECTED, 1);
+	ERR_FAIL_COND_V(incoming_packets.size() == 0, 1);
+
+	return incoming_packets.front()->get().from;
+}
+
+bool NetworkedMultiplayerCustom::is_server() const {
+	return self_id == 1;
+}
+
+void NetworkedMultiplayerCustom::poll() {
+}
+
+int NetworkedMultiplayerCustom::get_unique_id() const {
+	return self_id;
+}
+
+void NetworkedMultiplayerCustom::set_refuse_new_connections(bool p_refuse_new_connections) {
+	refusing_new_connections = p_refuse_new_connections;
+}
+
+bool NetworkedMultiplayerCustom::is_refusing_new_connections() const {
+	return refusing_new_connections;
+}
+
+NetworkedMultiplayerPeer::ConnectionStatus NetworkedMultiplayerCustom::get_connection_status() const {
+	return connection_status;
+}
+
+//
+// Custom methods.
+//
+
+void NetworkedMultiplayerCustom::initialize(int p_self_id) {
+	if (connection_status != ConnectionStatus::CONNECTION_CONNECTING) {
+		return;
+	}
+
+	self_id = p_self_id;
+	if (self_id == 1) {
+		set_connection_status(ConnectionStatus::CONNECTION_CONNECTED);
+	}
+}
+
+void NetworkedMultiplayerCustom::set_max_packet_size(int p_max_packet_size) {
+	max_packet_size = p_max_packet_size;
+}
+
+void NetworkedMultiplayerCustom::set_connection_status(NetworkedMultiplayerPeer::ConnectionStatus p_connection_status) {
+	// Can only go to connecting, if we are disconnected.
+	if (p_connection_status == ConnectionStatus::CONNECTION_CONNECTING && connection_status == ConnectionStatus::CONNECTION_DISCONNECTED) {
+		connection_status = p_connection_status;
+	}
+	// Can only go to connected, if we are connecting.
```

Comment:
I was playing around with this today and got hung up on this bit. An error message or some actual documentation on this restriction would be good.

Reply:
Thanks! I've created a new PR adding those: https://github.com/godotengine/godot/pull/63628